Add experimental "Staging Mode" for holding changes locally to be either committed all at once, or discarded

markfields opened the pull request:
## Description

Fixes [AB#34399](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/34399)

This PR introduces a new "Staging Mode" API to the runtime. When in Staging Mode, local changes are "Staged" and we don't try to submit them to the ordering service at all.  You can exit staging mode by either "discarding" or "committing" the staged changes.

**Discard** is similar to rollback, but more complex since remote changes that have been merged with the local changes need to be preserved even as the local changes are reverted.

**Commit** simply rebases and submits the local changes.

Note that this feature is experimental, and not fully implemented across all DDSes with this change.  Specifically:

* Rollback needs to be properly implemented on each DDS
* Rebasing before committing needs to support "squash rebase", where content that is added then removed is not included in the final set of ops sent over the wire.

## Reviewer Guidance

This is the final state of a long-lived branch.  You can browse the history of interesting changes by reviewing PRs targeting this branch:   https://github.com/microsoft/FluidFramework/pulls?q=is%3Apr+base%3Atest%2Fstaging-mode

markfields committed:
[test/staging-mode] Initial prototype (#23782)

Merging the initial prototype into the test branch for "Staging Mode".

---------

Co-authored-by: Tony Murphy <anthonm@microsoft.com>

markfields committed:
Merge branch 'main' into test/staging-mode

anthony-murphy committed:
test/staging-mode: Remove unneeded changes and enable testing (#23791)

Just cleaning up the test branch a bit by removing some speculative
changes that are not necessary, and trying to enable testing

anthony-murphy committed:
Merge branch 'main' into test/staging-mode

markfields committed:
[test/staging-mode] Addressing comments (#23786)

## Description

Address PR comments from initial prototype PR #23782 

## Reviewer Guidance

_not targeting `main` branch_

I use `//*` comments to indicate stuff to come back to before completing
the work. Bringing that convention to this branch, indicating it's
something to deal with before we merge any of this into `main`

---------

Co-authored-by: Tony Murphy <anthonm@microsoft.com>
Co-authored-by: Tony Murphy <anthony.murphy@microsoft.com>

anthony-murphy committed:
Merge branch 'main' into test/staging-mode

ChumpChief committed:
Merge branch 'main' into test/staging-mode

ChumpChief committed:
Merge branch 'main' into test/staging-mode

anthony-murphy committed:
Merge branch 'main' into test/staging-mode

anthony-murphy committed:
Merge branch 'main' into test/staging-mode

anthony-murphy committed:
Merge branch 'main' into test/staging-mode

anthony-murphy committed:
Merge branch 'main' into test/staging-mode

anthony-murphy committed:
Integrate Staging Mode into the Local Sever Stress Harness (#23969)

- Adds basic observability to staging mode to runtime and data stores
- Integrates staging mode in the Local Sever Stress Harness
- Allow the harness to replay multiple seeds, like only and skip

---------

Co-authored-by: Matt Rakow <ChumpChief@users.noreply.github.com>

anthony-murphy committed:
Staging Mode: Defer DDS and DataStore Attaches and Fix reconnecting while in staging mode bug (#23975)

This change fixes a number of issues that are hit by the stress tests:

1. Fixed can't rollback attach - We now defer attach to commit of
staging mode
2. Fixed out of order message due to reconnect - We now defer
reconnecting util staging mode is exited. This is a stop gap solution
for the prototype to avoid out of order messages.

---------

Co-authored-by: Matt Rakow <ChumpChief@users.noreply.github.com>

anthony-murphy committed:
Merge branch 'main' into test/staging-mode

anthony-murphy committed:
Staging Mode: Fix bind on resubmit (#24006)

some dds use the pre-serialized message on resubmit, this change catches
that and ensure the real handle is bound before sending the messages.
There is definitely a memory leak here in the reject case, which is
acceptable for the prototype, but we'll need a more robust solution once
we build the real thing.

anthony-murphy commented on packages/runtime/container-runtime/src/opLifecycle/opGroupingManager.ts:
```diff
@@ -159,8 +159,11 @@ export class OpGroupingManager {
 			// or be empty (to allow for empty batches to be grouped)
-			batch.messages.length !== 1
-			// Support for reentrant batches will be on by default
+			// TODO: Can we remove this, as it creates problems for staging mode
+			// as we always want re-submit, even if only 1 op.
+			(batch.messages.length !== 1 ||
+				// Support for reentrant batches will be on by default
+				batch.hasReentrantOps === true)
```
```suggestion
			// TODO: Can we remove this, as it creates problems for staging mode
			// as we always want re-submit, even if only 1 op.
			// i'm actually pretty sure there is a bug here.
			// without this change we will not rebase re-entrant ops
			// unless there is more than one op in the batch,
			// but even 1 op must be rebased, as it could be based
			// off a partial state as it was created reentrantly
			// while applying another batch
			(batch.messages.length !== 1 ||
			// Support for reentrant batches will be on by default
			batch.hasReentrantOps === true)
```

anthony-murphy committed:
Merge branch 'main' of https://github.com/microsoft/FluidFramework into test/staging-mode

markfields committed:
Merge branch 'main' into test/staging-mode

anthony-murphy committed:
Update packages/runtime/container-runtime/src/opLifecycle/opGroupingManager.ts

anthony-murphy committed:
Merge branch 'main' of https://github.com/microsoft/FluidFramework into test/staging-mode

anthony-murphy committed:
fix merge

markfields committed:
test/staging-mode: Hold staged changes in PSM not Outbox (#24051)

Until now, the prototype has held all staged changes in a single batch
in the Outbox. This has required some awkward workarounds, indicating
it's not the most ergonomic way to model staged changes.

This PR switches it so that staged changes are allowed through to the
Pending State Manager, but marked so that they are never actually
submitted.

Upon exiting Staging Mode:
- If committing, the staged flag is removed from all pending state, and
(if we're connected) those changes are replayed. (if disconnected,
they'll be resubmitted upon connect)
- If discarding, we will rollback each pending message similar to how
Outbox's rollback works

This will provide the same functionality as the current prototype but in
a way that is much more ergonomic and maintainable.

---------

Co-authored-by: Tony Murphy <anthony.murphy@microsoft.com>

anthony-murphy committed:
StagingMode: Move to experimental API (#24056)

Makes staging mode an experimental api. Overall, it feels a bit clunky
to use, I wonder if there is a better way to do it. I can think of some
other ways we could change how the new interfaces are exported, but it
wouldn't really change the bad things about the pattern

anthony-murphy committed:
Merge branch 'main' of https://github.com/microsoft/FluidFramework into test/staging-mode

anthony-murphy committed:
Merge branch 'main' of https://github.com/microsoft/FluidFramework into test/staging-mode

anthony-murphy committed:
Merge branch 'main' of https://github.com/microsoft/FluidFramework into test/staging-mode

anthony-murphy committed:
Merge branch 'main' of https://github.com/microsoft/FluidFramework into test/staging-mode

anthony-murphy committed:
update failing seeds

anthony-murphy committed:
Merge branch 'main' of https://github.com/microsoft/FluidFramework into test/staging-mode

anthony-murphy committed:
Merge branch 'main' of https://github.com/microsoft/FluidFramework into test/staging-mode

anthony-murphy committed:
Staging Mode: Integrate OrderSequentially with staging mode (#24124)

This change integrates staging mode into OrderSequentially which adds
support of attaches, and naively implements keeping the dirty state
consistent within staging mode it self.


[AB#34238](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/34238)

anthony-murphy committed:
Merge branch 'main' of https://github.com/microsoft/FluidFramework into test/staging-mode

markfields committed:
Exiting staging mode - final flush before flipping the bit (#24148)

The intent of the final flush is that any changes submitted during
staging mode should happen in staging mode, then we exit cleanly.

markfields committed:
Merge remote-tracking branch 'origin/main' into test/staging-mode

anthony-murphy committed:
merge fixes

markfields committed:
[Staging Mode] Remove wrong assert (#24367)

I added this assert during merging with main, shouldn't have done
that... 🙃

markfields committed:
fix unit test

anthony-murphy committed:
Merge branch 'main' of https://github.com/microsoft/FluidFramework into test/staging-mode

anthony-murphy committed:
update skips post merge

markfields committed:
Merge remote-tracking branch 'origin/main' into test/staging-mode

markfields committed:
Build fixes post-merge

markfields committed:
Merge remote-tracking branch 'origin/main' into test/staging-mode

markfields committed:
[Staging Mode] Revert Deferred Handles code (#24398)

Now that we are retaining the original runtime op in Pending State
Manager (See #24364), we no longer need to track deferred handles in
Serializer.

Additionally we can move the Staging Mode check into the implementation
of `IFluidHandle.bind`, removing it from the serializer code.

markfields committed:
[Staging Mode] Cleanup and polish in preparation for merging test branch into main (#24436)

No major behavior change intended here. Various cleanup and polish
changes, such as:

- Reverting unnecessary behavior changes
  - Back to Outbox rebase being skipped for single-message batches
- Reverting inconsequential diffs (renames, comments, etc)
- Nits like adding doc comments
- Adding some `//*` todo comments to resolve in subsequent PRs before
finalizing the branch for main
- Some cleanup in stagingMode.spec.ts

markfields committed:
[Staging Mode] Reenable type validation (no breaks!) (#24437)

To reduce friction we disabled type validation for several packages in
the test branch. We don't have any breaks anymore, so reenable.

anthony-murphy committed:
Merge branch 'main' of https://github.com/microsoft/FluidFramework into test/staging-mode

anthony-murphy committed:
Merge branch 'main' into test/staging-mode

Abe27342 commented on packages/runtime/container-runtime/src/containerRuntime.ts:
```diff
@@ -3253,6 +3270,68 @@ export class ContainerRuntime
+				if (this.connected) {
+					this.pendingStateManager.replayPendingStates(true /* onlyStagedBatched */);
+				} else {
+					this.pendingStateManager.clearStagingFlags();
```
Looking more deeply at how squashing integrates, and is there any deeper reason why we need to avoid calling `replayPendingStates` here when already disconnected? Once squash is supported, we will want to replay pending states regardless of current connection state so that we can squash them.

markfields replied:
Hmm good point. This depends on the resubmit that will happen on reconnect. But you're right we'll need to rebase/squash now. Let's make that change as a follow up though. Trying to converge this.

markfields commented on packages/runtime/container-runtime/src/containerRuntime.ts:
```diff
@@ -4263,6 +4342,10 @@ export class ContainerRuntime
+					//* Need tests for this - is it an ok strategy?
+					// Note: For now, we will never stage ID Allocation messages.
+					// They won't contain personal info and no harm in extra allocations in case of discarding the staged changes
+					staged: false,
```
@Abe27342 this is the ID Compressor thing. What happens if "unused" ID Allocation ops are submitted?

Abe27342 replied:
Nothing bad per se. The reconnect strategy id compressor uses is basically this. That said, since it's a little unorthodox, it might be worth chatting about. I'll scrutinize some of the relevant code paths on this PR.

Abe27342 commented on packages/runtime/container-runtime/src/containerRuntime.ts:
```diff
@@ -3253,6 +3270,68 @@ export class ContainerRuntime
+			//* NIT: Don't need to pass true.  And then these two params can be joined into an atomic object?
+			this.outbox.flush(undefined, true /* staged */);
+
+			this.stageControls = undefined;
```
(I know I'm commenting on the draft/WIP PR, so feel free to resolve this wherever makes sense) nit: seems better to clear this after invoking `discardOrCommit()`, maybe even in a try-finally.

markfields replied:
(This is actually very close to final)

Have to clear it first so the ops submitted during resubmit during commit don't get marked as staged

anthony-murphy replied:
we can continue to iterate here once the core is checked in. i've been refraining from doing to much work in this branch to avoid any thrashing while we move to main.

Abe27342 replied:
ah yeah of course

markfields committed:
[Staging Mode] Addressing final cleanup tasks (#24452)

Final pass over the `test/staging-mode` branch before flipping #23783
over for review.

markfields committed:
Merge remote-tracking branch 'origin/main' into test/staging-mode

markfields commented on packages/test/local-server-stress-tests/src/baseModel.ts:
```diff
@@ -69,6 +71,25 @@ export function makeGenerator<T extends BaseOperation>(
+				type: "exitStagingMode",
+				commit: random.bool(),
+			}),
+			25,
```
I wonder if we want to make this lower to get more changes in Staging Mode. At least in the minimized result JSON, it's often only 1 or 2 operations done in there.

anthony-murphy replied:
i actually think this work pretty well. we hit issue across all major op producing subsystem, attach, blobs, dds. i'm also not sure more ops is better.

something i've thought about is how do we validate the stress test itself. an idea i've had is we allow some kind of custom sampling/aggregation based on operations, and then have a validation phase at the end. this would allow tracking and checking anything. it might be a bit brittle, so need to be careful with it.

markfields committed:
Reverting unneeded changes

Abe27342 reviewed (approved):
This looks good enough to me to check it in and iterate--awesome!

w.r.t id-compressor, I think there are some lingering bugs with the simple strategy we've taken. I'm going to try and augment the suite in packages\test\test-end-to-end-tests\src\test\idCompressor.spec.ts to repro + add some coverage in this realm.
